refactors handling of parent types

xeno-by opened the pull request:
At the moment parser does too much w.r.t handling of parent types.
It checks whether a parent can have value arguments or not and
more importantly, it synthesizes constructors and super calls.

This approach is fundamentally incompatible with upcoming type macros.
Take for example the following two snippets of code:

  `class C extends A(2)`
  `class D extends A(2) with B(3)`

In the first snippet, `A` might be a type macro, therefore the super call
`A.super(2)` eagerly emitted by the parser might be meaningless. In the
second snippet parser will report an error despite that `B` might be
a type macro which expands into a trait.

Unfortunately we cannot simply augment the parser with the `isTypeMacro`
check. This is because to find out whether an identifier refers to a type
macro, one needs to perform a typecheck, which the parser cannot do.

Therefore we need a deep change in how parent types and constructors
are processed by the compiler, which is implemented in this commit.

xeno-by committed:
refactors handling of parent types

At the moment parser does too much w.r.t handling of parent types.
It checks whether a parent can have value arguments or not and
more importantly, it synthesizes constructors and super calls.

This approach is fundamentally incompatible with upcoming type macros.
Take for example the following two snippets of code:

  `class C extends A(2)`
  `class D extends A(2) with B(3)`

In the first snippet, `A` might be a type macro, therefore the super call
`A.super(2)` eagerly emitted by the parser might be meaningless. In the
second snippet parser will report an error despite that `B` might be
a type macro which expands into a trait.

Unfortunately we cannot simply augment the parser with the `isTypeMacro`
check. This is because to find out whether an identifier refers to a type
macro, one needs to perform a typecheck, which the parser cannot do.

Therefore we need a deep change in how parent types and constructors
are processed by the compiler, which is implemented in this commit.

xeno-by commented:
review @paulp @adriaanm @retronym 

This is a beta version of the pull request I'm going to submit next week after looking into it with Martin. It's quite polished already, so I decided to put it up for discussion.

I'm pretty much confident in the refactored parsing and typechecking logic, but I don't know much about positions. @dragos, if you could take a look, I would be very grateful. If you scroll all the way down, there's an `-Xprint-pos` printout that illustrates how positions are assigned in the new scheme of things.

Finally the pull request touches a bit of the error reporting logic. I'm just calling `setError` for indicate some typecheck errors affecting templates + I've traded the "inherits self" error for the "cyclic reference" error due to the reasons explained in the comment to `parentTypes` and `typedParentType`. @hubertp, I hope I didn't do anything blatantly incorrect.

paulp commented on src/reflect/scala/reflect/internal/TreeInfo.scala:
```diff
@@ -469,8 +455,34 @@ abstract class TreeInfo {
+   *  This extractor disassembles applications into the core, type arguments and list of lists of value arguments.
+   *  If the tree is not an application, it will be returned as-is with targs and vargss set to Nil.
+   */
+  object Application {
```
Let's find a better name for this than Application. Maybe something like Decompose? Elsewhere (I could swear I'd checked it in, but maybe not) I have this approximately identically as "PolyCall" with "Call" as a no-type-args 2-tuple form. But I realize "Call" is not especially accurate.

paulp replied:
Also, this would be a wonderful place not to do this:

```
Option[(Tree, List[Tree], List[List[Tree]])]
```

Instead,

```
case class Applied(sel: Tree, targs: List[Tree], argss: ArgLists) { }
Option[Applied]
```

That would be a big improvement even if you stopped right there; but at least as useful is the bonus of being able to write methods on Applied (or pick a name) which do the more common of the various select/apply/params/args manipulations which we are endlessly reimplementing.

xeno-by replied:
That sounds good. Then I could move `unwrapApply` here (along with a bunch of other methods - I guess, it'd be okay to pay the small performance price to stay clean). I like the name too - thanks.

paulp replied:
It's not clear to me that there's any performance cost at all. Actually I'll make a stronger statement and say this figures to be faster, because decomposing the data and having it all in one place will create easy opportunities to avoid re-performing the same work. It's not like an allocation of "case class Tuple3" is any cheaper than "case class Applied", and Applied also wins by virtue of offering exact argument types rather than T1, T2, T3. The couple of kilobyte bytecode cost of having an additional class is lost in the noise.

paulp replied:
If only extractors didn't put a gun to our head and make us use Option, we could do significantly better returning Applied and having a NoApplied object representing None. Maybe someday.

xeno-by replied:
I agree about `Applied` vs `Tuple3`. I meant the performance cost of eliminating `methPart` and replacing its usages with `Applied`, but probably it'd be negligible.

paulp commented on src/compiler/scala/tools/nsc/typechecker/StdAttachments.scala:
```diff
@@ -4,7 +4,32 @@ package typechecker
+      tree.attachments.get[SuperCallArgsAttachment] match {
+        case Some(SuperCallArgsAttachment(vargss)) => Some(vargss)
+        case _ => None
+      }
```
You can write this less noisily as

```
def unapply(tree: Tree) = tree.attachments.get[SuperCallArgsAttachment] collect { case SuperCallArgsAttachment(vargss) => vargss }
```

xeno-by replied:
Done

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1015/

paulp commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1484,125 +1477,244 @@ trait Typers extends Modes with Adaptations with Tags {
+      probe.initialize
+
+      if (probe.isTrait || inMixinPosition) {
+        // TODO: fix `templateParents` so that it doesn't mix up Nils with ListOfNils
```
This is important and it would be great to see it promoted from "TODO" to "DO".

I suggested years ago that we should have dedicated classes for the major List[List[T]] players (argss, paramss at least) rather than blurring the lines by shoehorning everything into a shared and lossy structure - lossy at least in the sense that Nil willingly hops between levels, and also because at the type level every List[List[Tree]] looks the same. The passage of time has only made some structure seem like a better idea.

xeno-by replied:
I'd say this is a bit ambitious for this PR, but we definitely need to discuss and tackle this before reflection gets frozen. It's good that we have a chance to meet next week.

Though fixing the immediate problem of confusing Nils and ListOfNils for just templates seems doable. I'll look into that tomorrow.

xeno-by replied:
`templateParents` is now fixed, so that we can distinguish `T` and `T()` in parent type lists. Now there's an interesting question. Should the following be an error? (It's not an error at the moment)

```
trait T
class C extends T()
```

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1015/

paulp commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1484,125 +1477,244 @@ trait Typers extends Modes with Adaptations with Tags {
+     *  Note the Select(Super(_, _), nme.CONSTRUCTOR) part. This is the representation of
+     *  a fill-me-in-later supercall dummy. The argss are Nil, which symbolizes the fact
+     *  that supercall argss are unknown during parsing and need to be transplanted from one of the parent types.
+     *  Read more about why the argss are unknown in `tools.nsc.ast.Trees.Template`.
```
It's a stretch to say that argss == Nil can convey such a specific sentiment, since every Nil looks like every other Nil. Can you expand this comment to clarify:
- what would it symbolize if argss != Nil?
- If the whole Select part is a dummy placeholder, why does it have to be so large? Is any of "Select(Super(This(tpnme.EMPTY), tpnme.EMPTY), nme.CONSTRUCTOR))" meaningful? By which I mean, are there conditions where the supercall dummy tree differs in any way from that one, and if so, where does it differ and what are the conditions?

My interest is to not drag around this laborious encoding of the supercall. It has been horrible dragging it around starting in the parser; pushing it out of the parser is progress, but not as much as would be a more direct encoding.

xeno-by replied:
The entire `Apply(Select(Super(This(tpnme.EMPTY), tpnme.EMPTY), nme.CONSTRUCTOR)), List())` is a dummy, and it's the one and only possible representation that can be emitted by parser. 

This tree is quite convenient though, because:
- It works as is for the case when no processing is required (empty ctor args for the superclass)
- Stripping off the `Apply` produces a core that only needs rewrapping with applications of actual argss.

For some time I was thinking of using just `Select(Super(This(tpnme.EMPTY), tpnme.EMPTY), nme.CONSTRUCTOR))`, but that one required wrapping even if the superclass doesn't take any argss.

xeno-by replied:
It would be great to have a dedicated `preSuper` field in `Template` and to synthesize the primary ctor on the fly in `typedTemplate`, but I'm not sure how this would work with namer. 

I suggest I explore this in 2.10.2. That'll be the time for macro annotations, and then I'm planning to dig into symbol tables and synthesis of methods and companions. Currently I'm under a mild pressure of delivering type macros soon, so I wouldn't go into this right now.

P.S. Can we freely change the signature of `Template`? Reflection is officially in flux, but by doing so we're going to screw up all the plugin writers, probably, the IDE and, most likely, SBT. What do you think?

paulp replied:
In that case, why not handle it more like emptyValDef, which is defined as

```
  object emptyValDef extends ValDef(Modifiers(PRIVATE), nme.WILDCARD, TypeTree(NoType), EmptyTree) {
    override def isEmpty = true
    super.setPos(NoPosition)
    override def setPos(pos: Position) = { assert(false); this }
  }
```

Make an object pendingSuperCall or something extend Apply(Select(Super(This(tpnme.EMPTY), tpnme.EMPTY), nme.CONSTRUCTOR)), List()). Maybe a little special handling in TreeCopiers is necessary (although emptyValDef seems to be getting by without it) but I think it would easily be worth it not to be constantly shuffling those meaningless nodes around. Given the amount of custom code involving the super-call, the least we can do is give it some identity beyond an agglomeration of application nodes and tpnme.EMPTYs.

xeno-by replied:
Oh that's a great idea! And I'll be able to put the logic of rewrapping the dummy with real args right into that object. Very nice.

paulp replied:
"Can we freely change the signature of Template?" As far as I'm concerned we can, and that and a quarter might buy you a stick of gum. Since none of my suggestions for changes to Trees were incorporated, I consider them to be rife with relatively easily fixable flaws. I won't be the one saying "oh well, we missed our chance, let's live with them as-is forever."

xeno-by replied:
http://groups.google.com/group/scala-internals/browse_thread/thread/e3507f461d0fac11

xeno-by replied:
On a second thought, unfortunately we won't be able make this another `emptyValDef`, because the `Super` part is supposed to get attributed during `typedPrimaryConstrBody`. 

We could introduce another attachment for that or change `SuperCallArgsAttachment` to accommodate for the attributed `Super`, and then using the attached info to adjust the primary constructor during `typedTemplate`. However, given the scope of necessary changes (beyond a few lines) and the fact that, according to Martin, the whole thing is to be rewritten soon, I'd say we don't do the follow-up refactoring.

xeno-by replied:
Disregard my previous comment

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1726/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1726/

xeno-by committed:
unifies approaches to call analysis in TreeInfo

Instead of a flurry of methods such as `methPart` or `typeArguments`
to analyze applications, there is a single access point to related info:

  `treeInfo.dissectApplied(tree)`

Dissection returns an instance of the `Applied` class, which can
extract parts of applications and perform other application-specific
operations, e.g. `applyCount`.

For the sake of convenience, there's also an extractor named `Applied`,
which extracts `core`, `targs` and `argss`, where core is everything
except targs and argss. Extractor works for both `Tree` and `Applied`
(in the former case, it's equivalent to first dissecting and pattern
matching).

xeno-by commented:
@paulp I've just introduced the `Applied` class, which is quite similar to what we discussed. Similar, but not quite it, because after looking into usages I decided to change some aspects. See the diff for examples - I think that affected code looks better now.

1) There's still an extractor for the common case, when we need to know core (aka methPart), targs or argss. Usage sites of former treeInfo methods suggested that it's a valuable thing to have.

2) Since the `Tree => Option[(...)]` covers typical use cases, getting to an instance of `Applied` goes through another route - the `treeInfo.dissectApplied` method.

3) Both extraction and dissection always succeed, i.e. even `Ident(foo)` can be extracted/dissected (in that case, core will be Ident(foo), whereas targs and argss will be Nils). This might look misleading at first, but initial inconvenience is, in my opinion, trumped by the conciseness of the code. Probably, `Applied` just needs a better name, which emphasizes its optionality.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1017/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1017/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1728/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1728/

xeno-by commented:
PLS REBUILD ALL

xeno-by commented:
...this comment is now obsolete...

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1729/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1729/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1018/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1018/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1730/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1730/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1019/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1019/

xeno-by commented:
Duh so much for the "beta" version of the pull request. Will resubmit later, when I handle all the errors.

retronym commented:
btw, you can also submit a PR to `2.10.x` on your own repo and invite previewers to comment over there. Not that this one was a problem, but in case you don't want to contribute to the occasionally over-full `scala/scala` queue.

xeno-by commented:
that's a nice idea, thanks! `scala/scala` has a benefit of an automatic rangepos build though.

xeno-by committed:
introduces global.pendingSuperCall

Similarly to global.emptyValDef, which is a dummy that stands for an
empty self-type, this commit introduces global.pendingSuperCall, which
stands for a yet-to-be-filled-in call to a superclass constructor.

pendingSuperCall is emitted by Parsers.template, treated specially by
Typers.typedParentType and replaced with a real superclass ctor call
by Typers.typedTemplate.

To avoid copy/paste, this commit also factors out and unifies dumminess
of EmptyTree, emptyValDef and pendingSuperCall - they all don't have a
position and actively refuse to gain one, same story for tpe.

xeno-by commented:
PLS REBUILD ALL

xeno-by commented:
Reopened after a bunch of fixes and introduction of global.pendingSuperCall. /cc @paulp 

While fine-tuning range positions I came up with the following solution, which seems to not violate any of the position invariants. Note how super argss migrate from `Template.parents` in parser to `<init>` in typer.

```
00:53 ~/Projects/Kepler_pre-typemacros/sandbox (topic/pre-typemacros)$ cat Test.scala
class C(x: Int, y: Int)
class D(x: Int, y: Int) extends C(x, y)
00:53 ~/Projects/Kepler_pre-typemacros/sandbox (topic/pre-typemacros)$ scalac -Xprint:parser,typer -Xprint-pos -Yrangepos Test.scala
[[syntax trees at end of                    parser]] // Test.scala
[0:63]package [0:0]<empty> {
  [0:23]class C extends [7:23][24]scala.AnyRef {
    [8:14]<paramaccessor> private[this] val x: [11:14]Int = _;
    [16:22]<paramaccessor> private[this] val y: [19:22]Int = _;
    [7]def <init>([8]x: [11]Int, [16]y: [19]Int) = [7]{
      [NoPosition][NoPosition][NoPosition]super.<init>();
      [7]()
    }
  };
  [24:63]class D extends [31:63][56:63][56:57]C([58:59]x, [61:62]y) {
    [32:38]<paramaccessor> private[this] val x: [35:38]Int = _;
    [40:46]<paramaccessor> private[this] val y: [43:46]Int = _;
    [31]def <init>([32]x: [35]Int, [40]y: [43]Int) = [31]{
      [NoPosition][NoPosition][NoPosition]super.<init>();
      [31]()
    }
  }
}

[[syntax trees at end of                     typer]] // Test.scala
[0:63]package [0:0]<empty> {
  [0:23]class C extends [7:23][24]scala.AnyRef {
    [8:14]<paramaccessor> private[this] val x: [8]Int = _;
    [16:22]<paramaccessor> private[this] val y: [16]Int = _;
    [24]def <init>([8]x: [11]<type: [11]scala.Int>, [16]y: [19]<type: [19]scala.Int>): [7]C = [24]{
      [24][24][24]C.super.<init>();
      [7]()
    }
  };
  [24:63]class D extends [31:63][56:57]C {
    [32:38]<paramaccessor> private[this] val x: [32]Int = _;
    [40:46]<paramaccessor> private[this] val y: [40]Int = _;
    [58:62]def <init>([32]x: [35]<type: [35]scala.Int>, [40]y: [43]<type: [43]scala.Int>): [31]D = [58:62]{
      [58:62][56][56]D.super.<init>([58:59]x, [61:62]y);
      [31]()
    }
  }
}
```

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1023/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1023/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1734/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1734/

xeno-by commented:
rebased against the latest 2.10.x

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1035/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1035/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1746/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1746/

xeno-by commented:
rebased against 2.10.x

@paulp @retronym please review

xeno-by commented:
PLS REBUILD ALL

hubertp commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1854,9 +1970,21 @@ trait Typers extends Modes with Adaptations with Tags {
+            deriveDefDef(primaryCtor)(block => Block(earlyVals :+ superCall, unit) setPos pos) setPos pos
+          case _ => primaryCtor
+        }
+        body mapConserve { case `primaryCtor` => primaryCtor1; case stat => stat }
```
This is the second run on body stats. Isn't it possible to move this inside the first flatmap?

xeno-by replied:
Ctor synthesis depends on `parents1` and `body`, which the namer doesn't know about. Also I wouldn't say that the primary ctor is derived from its template in the same sense that a getter is derived from a val, so including the logic into `addDerivedTrees` would be questionable. What do you think?

hubertp replied:
Sure, I wasn't suggesting putting it inside the addDerivedTrees but inside rewrappingWrapperTrees should be fine?

hubertp replied:
Anyway, that's how I would do it to avoid second run. I am not particularly attached to that piece of code so I will leave it up to you.

hubertp commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -2850,14 +2847,15 @@ self =>
-                      " may not have parameters", true)
+        if (in.token == LPAREN) {
+          if (parenMeansSyntaxError) syntaxError(s"traits or objects may not have parameters", true)
+          else assert(false, "unexpected opening parenthesis")
```
Didn't we prohibit assert(false, ...) ?

xeno-by replied:
What should be used instead?

hubertp replied:
A quick grep revealed that we either issue a warning (see GenASM) or unit.error (see patmat). Anything that doesn't scare the user is probably better than just having a false assert.

xeno-by replied:
If this branch is taken by the compiler, it indicates a fatal error, because this means I misunderstood how template parsing works. What do we do in that case? `abort`?

hubertp replied:
Probably.

paulp replied:
Yes, if one must hard abort "abort" is the way. Thank you hubert for picking up on my discouragement of assert-false.

xeno-by replied:
Done. Also evicted a dozen or so other usages of `assert(false)`.

hubertp commented on test/files/run/t5603.check:
```diff
@@ -12,7 +12,7 @@
     <119:139>def <init>([95]i: [98]Int) = <119:139>{
       <119:139>val nameElse = <134:139>"Bob";
-      [94][94][94]super.<init>();
+      [NoPosition][NoPosition][NoPosition]super.<init>();
```
Why do we have NoPosition here? It's because it is a dummy? Usually going to NoPosition is the way backwards (@dragos?).

xeno-by replied:
Yes, because it's a singleton dummy like `emptyValDef`. After typechecking the dummy is supposed to be replaced with a correctly positioned tree anyways.

hubertp replied:
Ah, ok. I didn't see that this output is from parser.

hubertp commented on test/files/neg/t877.check:
```diff
@@ -1,7 +1,7 @@
+t877.scala:3: error: ')' expected but eof found.
 trait Foo extends A(22A, Bug!) {}
-                   ^
+                                 ^
```
The new err msg here is confusing.

xeno-by replied:
This error is confusing I agree. However, it's not a disaster, since the preceding error clearly indicates a problem.

Moreover trunk is susceptible to the very same problem when parsing `class` definitions as described in https://issues.scala-lang.org/browse/SI-6763. 

This patch is a slight regression in a sense that it exposes `trait` definitions to the issue at hand, but I'd argue we can live with that given:
1) This is not a major error.
2) Along with the problematic error message compiler also emits another error message that can be used to get rid of this one.
3) We have a JIRA issue for the underlying problem.

paulp replied:
One should expect invalid literals to halt compilation before arriving at any more meaningful error messages, given that the program can't even be parsed. I don't think this is a problem.

paulp commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -948,12 +942,20 @@ trait Trees extends api.Trees { self: SymbolTable =>
   def ValDef(sym: Symbol): ValDef = ValDef(sym, EmptyTree)
 
-  object emptyValDef extends ValDef(Modifiers(PRIVATE), nme.WILDCARD, TypeTree(NoType), EmptyTree) {
+  trait DummyTree extends Tree {
```
Can this receive some documentation please? "Dummy" can mean all kinds of things, and elsewhere in the compiler doesn't mean what it does here (that is, "dummy type args", which are not placeholders the way that a pendingSuperCall is, nor is pendingSuperCall a distinguished object which exists only to satisfy the absence of the field like emptyValDef is.)

Plus as long as we're in here, "emptyValDef" is a horrible name. Let's call it "emptySelf".

It is not clear what isEmpty means on a Tree; it is not clear what isDummy means either.

xeno-by replied:
Created a separate issue for `emptyValDef`: https://issues.scala-lang.org/browse/SI-6762. Current situation with incompatible changes in scala-reflect.jar is unclear, so I won't include this refactoring in this pull request.

xeno-by replied:
Renamed `DummyTree` to `CanHaveAttrs`, renamed `isDummy` to `canHaveAttrs`, explained `isEmpty`. See the subsequent commits.

xeno-by committed:
evicts assert(false) from the compiler

xeno-by committed:
DummyTree => CannotHaveAttrs

This looks like a much more specific name than a generic "dummy" prefix.
CannotHaveAttrs also doesn't imply that an implementing tree should
satisfy some bigger contract. EmptyTree and emptyValDef are very unlike
each other, so there's no point in trying to unify them.

Also DummyTree.isEmpty is no longer automatically true. The notion of
trees not having positions and types by design (i.e. EmptyTree + empty
TypeTrees + emptyValDef + the new pendingSuperCall) is quite different
from the notion of an empty tree in a sense of a tree being a null object
(i.e. EmptyTree + empty TypeTrees).

xeno-by commented:
PLS REBUILD ALL

xeno-by commented:
PLS REBUILD ALL

xeno-by commented:
Addressed the feedback. Reviewers, please, take another look.

xeno-by committed:
more ListOfNil => Nil

Even more trees (together with Apply nodes produced by templateParents
and New nodes produced by New in TreeBuilders) now distinguish nullary
argument list from empty argument list.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1053/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1765/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1053/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1765/

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1083/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1083/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1794/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1794/

xeno-by commented:
The joy of rebasing. Will fix these errors soon.

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1800/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1800/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1090/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1090/

paulp commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -2742,20 +2742,17 @@ self =>
-      val parents = new ListBuffer[Tree] += startAnnotType()
-      val argss = (
-        // TODO: the insertion of ListOfNil here is where "new Foo" becomes
-        // indistinguishable from "new Foo()".
```
So does this mean I can now tell new Foo from new Foo() ?

xeno-by replied:
Yes. `typedParentType` and `internal.Trees.New` still silently convert Nil to ListOfNil, but it's their own choice, no longer imposed by the typer. Upd. typo: This should read "by the parser".

paulp replied:
Here is an example of why I care: this behavior is awful, and it cannot be fixed if these cases cannot be distinguished.

```
scala> class Foo[T](val x: T)
defined class Foo

scala> new Foo
res0: Foo[Unit] = Foo@6b285262

scala> new Foo()
res1: Foo[Unit] = Foo@7519d715

scala> new Foo(())
res2: Foo[Unit] = Foo@34bd86d4
```

paulp commented on src/compiler/scala/tools/nsc/typechecker/StdAttachments.scala:
```diff
@@ -4,7 +4,31 @@ package typechecker
+
+  /** Determines whether the given tree has an associated SuperCallArgsAttachment.
+   */
+  def carriesSuperCallArgs(tree: Tree): Boolean = superCallArgs(tree).nonEmpty
```
Can this be called "hasSuperCallArgs" to be consistent with dozens of methods called hasXXX (and none called carriesXXX.)

paulp replied:
Actually I prefer "hasSuperArgs", which can only mean the same.

paulp commented on src/reflect/scala/reflect/internal/pickling/UnPickler.scala:
```diff
@@ -309,7 +309,7 @@ abstract class UnPickler {
           else owner.newLinkedModule(clazz, pflags)
         case VALsym =>
-          if (isModuleRoot) { assert(false); NoSymbol }
+          if (isModuleRoot) { abort("") }
```
Let's set a good example for other aborters and come up with a message.

xeno-by replied:
Doh looks like I forgot to put a nice message here. Did that for ~5 other `assert(false)`, but forgot this one :(

xeno-by replied:
Done

paulp commented on src/reflect/scala/reflect/internal/util/Position.scala:
```diff
@@ -269,7 +270,7 @@ class OffsetPosition(override val source: SourceFile, override val point: Int) e
 class RangePosition(source: SourceFile, override val start: Int, point: Int, override val end: Int)
 extends OffsetPosition(source, point) {
-  if (start > end) assert(false, "bad position: "+show)
+  if (start > end) throw new FatalError("bad position: "+show)
```
Don't hand roll FatalError; call abort.

xeno-by replied:
Cannot call `abort` here - this location is outside the cake.

paulp replied:
Okay, then call sys.error.

xeno-by replied:
Done

paulp commented on src/reflect/scala/reflect/runtime/JavaMirrors.scala:
```diff
@@ -416,7 +416,7 @@ private[reflect] trait JavaMirrors extends internal.SymbolTable with api.JavaUni
           case sym if sym == Predef_classOf           => fail("Predef.classOf is a compile-time function")
           case sym if sym.isTermMacro                 => fail(s"${symbol.fullName} is a macro, i.e. a compile-time function")
-          case _                                      => assert(false, this)
+          case _                                      => abort("unsupported symbol $symbol when invoking $this")
```
Somebody forget their interpolator...

xeno-by replied:
Done

paulp commented:
This will lg to m after the handful of comments I left have been addressed.

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1097/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1807/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1097/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1807/

adriaanm commented:
LGTM
